Add cat forward operation

et16kr opened the pull request:
- Added cat operation and kernel.
- Added driver test and gtest for cat
-  Compared to ROCm pytorch, there was a performance improvement when the number of input tensors was limited to 8 or fewer and output tensor size under 1M. 
- Cat driver example
```
# ./MIOpenDriver catbfp16 -t 1 -d 1 --input1 4x136800x91 --input2 4x34200x91 --input3 4x8550x91 --input4 4x2223x91 --input5 4x630x91
MIOpenDriver catbfp16 -t 1 -d 1 --input1 4x136800x91 --input2 4x34200x91 --input3 4x8550x91 --input4 4x2223x91 --input5 4x630x91
PRNG seed: 12345678
GPU Kernel Time Forward Cat Elapsed: 0.323147 ms
Forward Cat Verifies OK on CPU reference
```
- Added solver for cat

<details><summary>float16</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">


op_name | dtype | size | dim | model | direction | ROCm pytorch | MIOpen | imporvement
-- | -- | -- | -- | -- | -- | -- | -- | --
cat | float16 | [2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1560128 | 1553621 | 1.00
cat | float16 | [32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 17600 | 9938 | 1.77
cat | float16 | [32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 39200 | 25084 | 1.56
cat | float16 | [16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 128496 | 94151 | 1.36
cat | float16 | [64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 40608 | 31787 | 1.28
cat | float16 | [65536,16],[65536,351] | 1 | dlrm | fwd | 213136 | 240872 | 0.88
cat | float16 | [512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 341232 | 243147 | 1.40
cat | float16 | [256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 197744 | 166098 | 1.19
cat | float16 | [256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 82672 | 63840 | 1.29
cat | float16 | [2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 35696 | 24569 | 1.45
cat | float16 | [16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 34432 | 26880 | 1.28
cat | float16 | [3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 304816 | 116640 | 2.61
cat | float16 | [64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 56144 | 31147 | 1.80
cat | float16 | [12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1064288 | 346365 | 3.07
cat | float16 | [6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 44928 | 35147 | 1.28
cat | float16 | [255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 14608 | 13138 | 1.11
cat | float16 | [4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 529008 | 328747 | 1.61
cat | float16 | [256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 54944 | 32782 | 1.68
cat | float16 | [256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 96912 | 65244 | 1.49
cat | float16 | [2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 46528 | 24765 | 1.88
cat | float16 | [16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 40464 | 22756 | 1.78
cat | float16 | [4096,1312],[4096,1312] | 0 | rnnt | fwd | 58608 | 49742 | 1.18
cat | float16 | [9600,384],[128,384] | 0 | roberta | fwd | 27744 | 18578 | 1.49
cat | float16 | [256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 1672624 | 1017851 | 1.64
cat | float16 | [16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 68304 | 47698 | 1.43
cat | float16 | [1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 981168 | 840339 | 1.17
cat | float16 | [724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2123039 | 1905106 | 1.11



</body>

</html>

</details>

<details><summary>float32</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">


op_name | dtype | size | dim | model | direction | ROCm pytorch | MIOpen | imporvement
-- | -- | -- | -- | -- | -- | -- | -- | --
cat | float32 | [2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1810432 | 1774453 | 1.02
cat | float32 | [32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 19648 | 11716 | 1.68
cat | float32 | [32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 46544 | 29582 | 1.57
cat | float32 | [16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 148640 | 112427 | 1.32
cat | float32 | [64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 90480 | 43556 | 2.08
cat | float32 | [65536,16],[65536,351] | 1 | dlrm | fwd | 280464 | 273067 | 1.03
cat | float32 | [512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 430912 | 338454 | 1.27
cat | float32 | [256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 297472 | 178524 | 1.67
cat | float32 | [256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 99600 | 71502 | 1.39
cat | float32 | [2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 46944 | 32053 | 1.46
cat | float32 | [16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 65824 | 35307 | 1.86
cat | float32 | [3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 534320 | 215094 | 2.48
cat | float32 | [64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 57824 | 38756 | 1.49
cat | float32 | [12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1663424 | 648694 | 2.56
cat | float32 | [6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 90416 | 44942 | 2.01
cat | float32 | [255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 23600 | 15218 | 1.55
cat | float32 | [4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 974207 | 525370 | 1.85
cat | float32 | [256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 59920 | 49653 | 1.21
cat | float32 | [256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 107760 | 97387 | 1.11
cat | float32 | [2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 51552 | 32391 | 1.59
cat | float32 | [16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 44464 | 29013 | 1.53
cat | float32 | [4096,1312],[4096,1312] | 0 | rnnt | fwd | 96160 | 66205 | 1.45
cat | float32 | [9600,384],[128,384] | 0 | roberta | fwd | 43536 | 25724 | 1.69
cat | float32 | [256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 2807327 | 1580038 | 1.78
cat | float32 | [16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 89424 | 58222 | 1.54
cat | float32 | [1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 1104256 | 900588 | 1.23
cat | float32 | [724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2332688 | 1915719 | 1.22



</body>

</html>


</details>

<details><summary>bfloat16</summary>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->

</head>

<body link="#0563C1" vlink="#954F72">


op_name | dtype | size | dim | model | direction | ROCm pytorch | MIOpen | imporvement
-- | -- | -- | -- | -- | -- | -- | -- | --
cat | bfloat16 | [2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1561328 | 1550011 | 1.01
cat | bfloat16 | [32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 17504 | 9956 | 1.76
cat | bfloat16 | [32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 39360 | 25067 | 1.57
cat | bfloat16 | [16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 148560 | 94951 | 1.56
cat | bfloat16 | [64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 41760 | 31484 | 1.33
cat | bfloat16 | [65536,16],[65536,351] | 1 | dlrm | fwd | 212672 | 240694 | 0.88
cat | bfloat16 | [512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 340272 | 242382 | 1.40
cat | bfloat16 | [256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 197952 | 166009 | 1.19
cat | bfloat16 | [256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 83264 | 64249 | 1.30
cat | bfloat16 | [2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 35904 | 23733 | 1.51
cat | bfloat16 | [16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 34144 | 27076 | 1.26
cat | bfloat16 | [3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 310192 | 116498 | 2.66
cat | bfloat16 | [64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 56192 | 31093 | 1.81
cat | bfloat16 | [12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1060896 | 345974 | 3.07
cat | bfloat16 | [6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 44672 | 35716 | 1.25
cat | bfloat16 | [255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 14528 | 13102 | 1.11
cat | bfloat16 | [4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 526272 | 327218 | 1.61
cat | bfloat16 | [256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 54240 | 32853 | 1.65
cat | bfloat16 | [256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 96800 | 65582 | 1.48
cat | bfloat16 | [2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 45392 | 24533 | 1.85
cat | bfloat16 | [16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 40256 | 22382 | 1.80
cat | bfloat16 | [4096,1312],[4096,1312] | 0 | rnnt | fwd | 57616 | 50027 | 1.15
cat | bfloat16 | [9600,384],[128,384] | 0 | roberta | fwd | 27568 | 18293 | 1.51
cat | bfloat16 | [256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 1672320 | 1002810 | 1.67
cat | bfloat16 | [16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 66848 | 48284 | 1.38
cat | bfloat16 | [1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 982031 | 841744 | 1.17
cat | bfloat16 | [724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2114175 | 1904376 | 1.11



</body>

</html>
</details>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{color:#1F2328;
	font-size:11.0pt;
	font-weight:700;
	font-family:"Segoe UI", sans-serif;
	mso-font-charset:0;
	text-align:center;
	white-space:normal;}
.xl66
	{color:#1F2328;
	font-size:11.0pt;
	font-family:"Segoe UI", sans-serif;
	mso-font-charset:0;
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">


type | direction | avrage
-- | -- | --
float16 | fwd | 1.51
float | fwd | 1.58
bfloat16 | fwd | 1.52



</body>

</html>

et16kr committed:
prototype op cat

et16kr committed:
add kernel for cat

et16kr committed:
add gtest for cat

et16kr committed:
add driver for cat

et16kr committed:
fix seg fault for cat unit test

et16kr committed:
fix cat driver

et16kr committed:
cppcheck for cat op and remove MIOPEN_BETA_API

et16kr committed:
add document for cat

et16kr committed:
fix value name inputDesc to xDesc for cat

et16kr committed:
remove check 0 size tensor for cat

et16kr committed:
add unit test case for cat

et16kr committed:
remove some comment and double dtype

et16kr committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
fix merge conflict for cat

et16kr committed:
mloCatHost move to cat_driver and update cat_driver error msg

junliume committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
in CatDriver and test_cat, CPU and GPU outputs must match exactly

et16kr committed:
update message for CatDriver

junliume committed:
Merge branch 'develop' into impl_cat_forward

JehandadKhan commented on include/miopen/miopen.h:
```diff
@@ -2531,6 +2532,35 @@ MIOPEN_EXPORT miopenStatus_t miopenLayerNormForward(miopenHandle_t handle,
+ * @return               miopenStatus_t
+ */
+MIOPEN_EXPORT miopenStatus_t miopenCatForward(miopenHandle_t handle,
+                                              const std::vector<miopenTensorDescriptor_t>& xDescs,
```
The API cannot have C++ constructs.

JehandadKhan replied:
Moreover, it would be a better design to use the existing MIOpen TensorOp API with a new enum for the op itself.

et16kr replied:
- Convert vector to pointer array and array size.
- The TensorOP is accepted only two input tensors, whereas the Cat requires a variable number of input tensors and a specified concatenation dimension. The formats of the two API differ.

JehandadKhan commented on src/kernels/MIOpenCat.cpp:
```diff
@@ -0,0 +1,145 @@
+                             FLOAT* __restrict__ output,
+                             const uint64_t input_dim_size,
+                             const uint64_t stride,
+                             uint64_t* output_offset)
```
```suggestion
template <typename T>
__device__ void cat_copy_buf(const T* __restrict__ input,
                             T* __restrict__ output,
                             const uint64_t input_dim_size,
                             const uint64_t stride,
                             uint64_t* output_offset)
```

JehandadKhan replied:
Also, 64-bit math is going to limit the capabilities of this kernel, I suggest we have two versions, one with 32-bit math and another with 64-bit math, and switch to the 64-bit version only if necessary.

et16kr replied:
- Transformed the function to template.
- I have converted "dim" and "stride" from int64_t to int, while changing "Offset" and "size" to size_t.
If the request wasn't about the offset variable, please let me know again.

et16kr committed:
cat Kernel function has been converted to a template

et16kr committed:
remove vector from miopenCatForward

et16kr committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
change vector to pointer array at CatForward args

et16kr committed:
prototype cat forward solver

et16kr committed:
Excluding cat unittest with slower performance than ROCm.

et16kr committed:
fix limit input tensor count for cat

et16kr committed:
Merge branch 'develop' into impl_cat_forward

junliume committed:
Merge branch 'develop' into impl_cat_forward

junliume committed:
Merge branch 'develop' into impl_cat_forward

CAHEK7 commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,253 @@
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 256
+#define MAX_TENSOR_X_COUNT 8
```
Why do we have such a restriction?

Moreover can it be a local `constexpr` value instead of global define?

atamazov replied:
> Moreover can it be a local `constexpr` value instead of global define?

The macro can't be global or local as it does not have linkage. In this specific case, the macro is visible in forward_cat.cpp only.

CAHEK7 replied:
I mean it is visible to every entity in the file.

Vsevolod1983 commented on driver/cat_driver.hpp:
```diff
@@ -0,0 +1,341 @@
+                             std::vector<Tgpu*> inputs,
+                             miopenTensorDescriptor_t outputDesc,
+                             Tcheck* outputhost,
+                             int32_t dim)
```
Should we use unsigned type for "dim" ? It can't be negative, right ?

et16kr replied:
change to uint32_t

Vsevolod1983 commented on driver/cat_driver.hpp:
```diff
@@ -0,0 +1,341 @@
+    InputFlags inflags;
+
+    int forw;
+    int dim_size;
```
Should :dim" and "dim_size" has unsigned type?

et16kr replied:
change to uint32_t too
and remove unused forw

Vsevolod1983 commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -0,0 +1,150 @@
+    int32_t GetDim() const { return dim; }
+    int32_t GetXCount() const { return xCount; }
+
+    bool IsSameType() const
```
Functions which return bool are intended to return true or false. Its not expected to get an exception in case if false value should be returned. It might be not safe from a caller's perspective.
Also, there are several functions which have same behavior below.

CAHEK7 commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,253 @@
+#define MAX_TENSOR_X_COUNT 8
+#define MIN_OUTPUT_TENSOR_SIZE 1000000
+
+namespace miopen {
+
+namespace solver {
+
+namespace cat {
```
A small suggestion. @atamazov @averinevg @JehandadKhan what do you think?

It's possible to squash namespaces in that way, but it's mostly a style change (also it can be forced by clang-format)
```suggestion
namespace miopen::solver::cat {
```
Proper fix is a clang-format rule, no need to do anything in this PR.

CAHEK7 commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,253 @@
+
+#define LOCAL_SIZE 256
+#define MAX_TENSOR_X_COUNT 8
+#define MIN_OUTPUT_TENSOR_SIZE 1000000
```
Can it be a local `constexpr` value instead of global define? It brings some more meaning to the var, since we can understand what kind of heuristic it is.

atamazov replied:
@CAHEK7  see https://github.com/ROCm/MIOpen/pull/2562#discussion_r1436156749

CAHEK7 replied:
Anyway, the meaning of `MIN_OUTPUT_TENSOR_SIZE` becomes clearer when it is placed right in the function `IsImprovementOverROCm` and it is the only place where we use it.
There is no need to keep this named constant accessible everywhere in the file.

et16kr replied:
I have changed the #define macro to a constexpr constant within the function.

et16kr replied:
I have changed the #define macro to a constexpr constant within the function.

CAHEK7 commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,253 @@
+bool CatForward::IsApplicable(const ExecutionContext& context,
+                              const miopen::cat::ProblemDescription& problem) const
+{
+    if(!IsUnderXCountLimit(problem))
+        return false;
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsRightDim())
+        return false;
+    if(!problem.IsRightLength())
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    if(!IsImprovementOverROCm(problem))
+        return false;
```
All those functions must not throw exceptions without a strong reason when we can't compute the result at all.
The meaning of `IsApplicable` is to say that we can't use this solver for that particular data. If the data is not suitable for the solver, we can just skip it returning `false`, instead of kind of crashing.

et16kr replied:
I removed exception, and set return false.

CAHEK7 commented on src/kernels/MIOpenCat.cpp:
```diff
@@ -0,0 +1,146 @@
+                             T* __restrict__ output,
+                             const size_t input_dim_size,
+                             const int stride,
+                             size_t* output_offset)
```
```suggestion
__device__ size_t cat_copy_buf(const T* __restrict__ input,
                             T* __restrict__ output,
                             const size_t input_dim_size,
                             const size_t stride,
                             size_t output_offset)
```
It looks much clearer and more explicit from API point of view. What do you think?

et16kr replied:
fixed it

CAHEK7 commented on src/kernels/MIOpenCat.cpp:
```diff
@@ -0,0 +1,146 @@
+    if(!input)
+        return;
+
+    int gid0            = blockIdx.x * blockDim.x + threadIdx.x;
+    int gid1            = blockIdx.y * blockDim.y + threadIdx.y;
+    int gsz0            = gridDim.x * blockDim.x;
+    size_t input_offset = static_cast<size_t>(gid1) * input_dim_size * static_cast<size_t>(stride);
```
```suggestion
    size_t gid0            = blockIdx.x * blockDim.x + threadIdx.x;
    size_t gid1            = blockIdx.y * blockDim.y + threadIdx.y;
    size_t gsz0            = gridDim.x * blockDim.x;
    size_t input_offset = gid1 * input_dim_size * stride;
```
All those variables has always been casted to `size_t`

et16kr replied:
change to size_t

CAHEK7 commented on src/kernels/MIOpenCat.cpp:
```diff
@@ -0,0 +1,146 @@
+                                         FLOAT* __restrict__ output,
+                                         const size_t input1_dim_size,
+                                         const size_t input2_dim_size,
+                                         const int dim,
```
`dim` is unused

et16kr replied:
I removed it.

CAHEK7 reviewed (changes requested):
Partial review.

Vsevolod1983 commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,253 @@
+{
+    if(!IsUnderXCountLimit(problem))
+        return false;
+    if(!problem.IsSameType())
```
IsSameType and some other functions  never returns false (it throws an exception instead). Is it intended?

atamazov replied:
@Vsevolod1983 Good catch! #2639 opened.

et16kr replied:
fix it, thank you.

atamazov replied:
[resolved]

atamazov commented:
@et16kr How urgent is this for you?

/cc @junliume

et16kr committed:
fix pr review, cat_copy_buf, int to size_t, return next offset, remove unused dim

et16kr committed:
fix cat pr review, change exception to return false, in cat

et16kr committed:
change int to unsigned, remove unused variable

et16kr commented:
> @et16kr How urgent is this for you?
> 
> /cc @junliume

It's not urgent. Thank you.

et16kr committed:
fix cat_copy_buf return offset when input is null

atamazov commented on driver/cat_driver.hpp:
```diff
@@ -0,0 +1,340 @@
+    {
+        auto input       = inputs[i];
+        size_t dim_size  = miopen::deref(inputDescs[i]).GetLengths()[dim];
+        size_t copy_size = inner_size / output_dim_size * dim_size;
```
[Q] Is it guaranteed that `inner_size % output_dim_size == 0`?

et16kr replied:
It was guaranteed that 'inner_size' includes 'output_dim_size', 
but I have modified the code to avoid confusion by separating 'output_dim_size' from 'inner_size'.

atamazov replied:
[resolved]

atamazov commented on driver/cat_driver.hpp:
```diff
@@ -0,0 +1,340 @@
+private:
+    InputFlags inflags;
+
+    uint32_t dim_size;
```
This member is unused in this PR.

et16kr replied:
remove it

atamazov replied:
[resolved]

atamazov reviewed (changes requested):
Partial

atamazov commented:
@junliume Maybe set https://github.com/ROCm/MIOpen/labels/urgency_normal?

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -0,0 +1,146 @@
+
+        if(ydims[dim] != yDesc.GetLengths()[dim])
+        {
+            return false;
```
@et16kr @Vsevolod1983 It was
```C++
MIOPEN_THROW(miopenStatusBadParm, "CatForward: Tensor dimension lengths do not match.");
```
which means that this `if` detects the case when PorblemConfig is incorrect.

❗ The constructor of ProblemConfig must fail when/if the configuration is invalid/incorrect.

Knowing that, we can select one of the following approaches in the rest of the code:
- (A) Ignore invalid configs (as we "assume" that invalid ProblemConfig can't be constructed).
  - ⚠️ But assumptions are generally bad. Each assumption is an implicit dependence.
    - What if an assumption becomes incorrect in the future? When using assumptions we have to always analyze such scenarios.
- (B.1) Detect and throw exception
  - This is good for builds with `DEV=On` and for the _debug_ builds.
- (B.2) Detect and handle in some other way, if possible (like this function does now).
  - This approach is good for the _release_ builds.

Given that, the best approach for _this specific_ `if` in this function is to `return false` for the release builds and to `throw` for the debug/DEV builds (however always returning `false` is also fine).

Please note that it is fine to ignore invalid ProblemConfigs in all other solver methods, because `IsApplicable()` is always called first. For example, we can safely ignore incorrect ProblemConfigs in `GetSolution()`.

Action points:
- [ ] (1) Check if ProblemConfig constructor fails when/if the configuration is invalid/incorrect. Fix it, if necessary.
- [ ] (2) Consider if you would like to use the following approach in `IsApplicable()`:
```C++
#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
    MIOPEN_THROW(...); // this is (B.1)
#else
    return false;  // this is (B.2)
#endif
```

The final decision about AP (2) is totally up to you.

/FYI @CAHEK7 @DrizztDoUrden @averinevg @shurale-nkn

et16kr replied:
I have modified the code as suggested.

atamazov replied:
@et16kr Unresolved. This is not implemented yet:

> ❗ The constructor of ProblemConfig must fail when/if the configuration is invalid/incorrect.

When the above is implemented, MIOPEN_THROW can be removed from member functions because in this specific case it is guaranteed that the invalid ProblemConfig object won't be contructed.

In short, the `MIOPEN_THROW` cosntructs may be simply moved into the contructor.

et16kr replied:
I have made changes to have the verification of dim and length in the constructor. 
Additionally, I removed exceptions from the IsAllPacked().

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -68,7 +68,7 @@ struct ProblemDescription : ProblemDescriptionBase
             if(xDescs[i]->GetType() != dtype)
             {
-                MIOPEN_THROW(miopenStatusBadParm, "CatForward: Tensor types do not match.");
+                return false;
```
Same as https://github.com/ROCm/MIOpen/pull/2562#discussion_r1436579150

et16kr replied:
I've modified it to return false in release and throw an exception with an error message in debug.

atamazov replied:
This `MIOPEN_THROW` should reside in the constructor. And you can remove it from here. See https://github.com/ROCm/MIOpen/pull/2562#discussion_r1437749430.

et16kr replied:
move it to constructor

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -84,24 +84,22 @@ struct ProblemDescription : ProblemDescriptionBase
             {
-                MIOPEN_THROW(miopenStatusBadParm,
-                             "CatForward: Tensor dimension lengths do not match.");
+                return false;
```
Ditto

et16kr replied:
changed it.

atamazov replied:
See https://github.com/ROCm/MIOpen/pull/2562#discussion_r1437751964

et16kr replied:
move it to constructor, too

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -84,24 +84,22 @@ struct ProblemDescription : ProblemDescriptionBase
                 {
-                    MIOPEN_THROW(miopenStatusBadParm,
-                                 "CatForward: Tensor dimension lengths do not match.");
+                    return false;
```
Ditto

et16kr replied:
changed it.

atamazov replied:
See https://github.com/ROCm/MIOpen/pull/2562#discussion_r1437751964

et16kr replied:
move it to constructor

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -111,9 +109,7 @@ struct ProblemDescription : ProblemDescriptionBase
-            MIOPEN_THROW(
-                miopenStatusBadParm,
-                "CatForward: is greater than 0 and less than or equal tensor dimension length.");
+            return false;
```
Ditto

et16kr replied:
changed it.

atamazov replied:
See https://github.com/ROCm/MIOpen/pull/2562#discussion_r1437751964

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -124,13 +120,13 @@ struct ProblemDescription : ProblemDescriptionBase
             if(!xDescs[i]->IsPacked())
             {
-                MIOPEN_THROW(miopenStatusBadParm, "CatForward: Unpacked tensors not supported.");
+                return false;
```
Ditto

et16kr replied:
changed it.

atamazov replied:
[resolved here - see next review]

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -124,13 +120,13 @@ struct ProblemDescription : ProblemDescriptionBase
         if(!yDesc.IsPacked())
         {
-            MIOPEN_THROW(miopenStatusBadParm, "CatForward: Unpacked tensors not supported.");
+            return false;
```
Ditto

et16kr replied:
changed it.

atamazov replied:
[resolved here - see next review]

et16kr replied:
remove exception, It returns false if it is not packed.

atamazov reviewed (changes requested):
Wrt AP (1). It seems like these checks are missing from `ProblemDescription()`

et16kr committed:
cat driver, remove unused value

et16kr committed:
add exceptions for debug, cat forward IsApplicable

et16kr committed:
change define macro to function constexpr value in cat forward

CAHEK7 commented on test/gtest/cat.hpp:
```diff
@@ -0,0 +1,198 @@
+    {
+        auto&& handle = get_handle();
+        cat_config    = GetParam();
+        std::mt19937 gen(0);
+        std::uniform_real_distribution<> d{-3, 3};
+        auto gen_value = [&](auto...) { return d(gen); };
```
Please use generators from driver/random.hpp and test/random.hpp

Here is an example:
https://github.com/ROCm/MIOpen/pull/2583/commits/b90bf90ffd259e2c5a78efa183cb7d73615e1827#diff-bfa18307b5c43a529901ac79b9428ea97493d6d80182903fd8fa843aa3d0386c

et16kr replied:
The test time has been reduced to 1/4.5. Thank you.

et16kr committed:
test_cat, change gen random value

junliume commented on src/include/miopen/cat.hpp:
```diff
@@ -0,0 +1,45 @@
+struct TensorDescriptor;
+
+miopenStatus_t CatForward(Handle& handle,
+                          const int32_t xCount,
```
@et16kr Found some hip tidy issues:
```
error: parameter 'xCount' is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions [readability-avoid-const-params-in-decls,-warnings-as-errors]
```

et16kr replied:
This const was removed, in [c17a70e].

junliume commented on src/solver/cat/forward_cat.cpp:
```diff
@@ -0,0 +1,249 @@
+    return true;
+}
+
+bool CatForward::IsApplicable(const ExecutionContext& context,
```
@et16kr make `context` `[[maybe_unused]] ` ?

et16kr replied:
This variable was already commented, you might have seen the code from the previous revision.
And now I changed it to [[maybe_unused]].

et16kr replied:
Yes, this function parameter was previously commented, in [c17a70e].
And I have now changed it to include the [[maybe_unused]] keyword.

junliume commented on src/include/miopen/cat/solvers.hpp:
```diff
@@ -0,0 +1,62 @@
+    std::size_t GetWorkspaceSize(const ExecutionContext& context,
+                                 const miopen::cat::ProblemDescription& problem) const override
+    {
+        return 0;
```
@et16kr is it a WIP? same return regardless of input?

et16kr replied:
Yes, cat don't use the workspace. So, it's a return of 0. 
And function parameters were previously commented, but I have now changed them to include the [[maybe_unused]] keyword.

junliume commented on src/include/miopen/cat/cat_invoke_params.hpp:
```diff
@@ -0,0 +1,66 @@
+                    const TensorDescriptor& yDesc_,
+                    void* y_,
+                    const int32_t dim_)
+        : xCount(xCount_), xDescs(xDescs_), xs(xs_), yDesc(yDesc_), y(y_), dim(dim_)
```
@et16kr we may need to add `clang-analyzer-optin.cplusplus.UninitializedObject` to `.clang-tidy` file, due to https://github.com/llvm/llvm-project/issues/61919 may not yet reached `amdclang`

et16kr replied:
init member variables

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -120,13 +143,21 @@ struct ProblemDescription : ProblemDescriptionBase
+                MIOPEN_THROW(miopenStatusBadParm, "CatForward: Unpacked tensors not supported.");
+#else
                 return false;
+#endif
```
Do not throw, this doesn't mean that ProblemConfig is invalid.
```suggestion
                return false;
```

et16kr replied:
remove exception, It returns false if it is not packed.

atamazov commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -120,13 +143,21 @@ struct ProblemDescription : ProblemDescriptionBase
         if(!yDesc.IsPacked())
         {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+            MIOPEN_THROW(miopenStatusBadParm, "CatForward: Unpacked tensors not supported.");
```
Ditto

et16kr replied:
remove exception, It returns false if it is not packed.

atamazov reviewed (changes requested):
`IsApplicable()` should not throw when ProblemConfig is valid.

et16kr committed:
remove exception from CatForward::IsApplicable()

et16kr committed:
remove unused function args identified by clang-tidy

et16kr committed:
Merge branch 'develop' into impl_cat_forward

JehandadKhan commented:
@atamazov and @CAHEK7 another ping for reviews.

JehandadKhan commented on docs/cat.rst:
```diff
@@ -0,0 +1,13 @@
+To enable this, define MIOPEN_BETA_API before including miopen.h.
+
+
+miopenCatForward
```
Perhaps a few lines describing what the operation is would be useful.

et16kr replied:
I added the following sentence to the cat doc.
```Concatenates the given sequence of seq tensors in the given dimension.```

JehandadKhan replied:
If possible, please add an example as well, that would help others to understand the use for the API.

et16kr replied:
I've made a  prototype for the sample code, (I haven't applied it to the doc yet.)
if kept short, it same the arguments list, 
and if expanded, most of the content pertains to tensor creation.
Shall I refine and apply the following content?
Or please let me know of any other op's examples for referencing.

```
For example: two tensors are concatenated to a single output tensor.

#include <miopen/miopen.h>

int main(int argc, char* argv[])
{
    const int input_count = 2;
    const int input_shape_size = 4;
    miopenHandle_t handle;
    int dim = 1;
    int input_shape[input_count][input_shape_size] = {{32,128,128,128}, {32,64,128,128}};
    int output_shape[input_shape_size] = {32,192,128,128};
    void* input_gpu_buf[2];
    float* input_buf[2];
    void* output_gpu_buf;
    float* output_buf;
    int input_size[2] = {1, 1};
    int output_size = 0;
    float* input0, *input1, *output;
    bool is_different = false;

    miopenTensorDescriptor_t inputDescs[input_count];
    miopenTensorDescriptor_t outputDesc;

    for (int i = 0 ; i < input_shape_size ; i++)
    {
        input_size[0] *= input_shape[0][i];
        input_size[1] *= input_shape[1][i];
    }
    output_size = input_size[0] + input_size[1];

    input_buf[0] = (float*)malloc(input_size[0] * sizeof(float));
    input_buf[1] = (float*)malloc(input_size[1] * sizeof(float));
    output_buf = (float*)malloc(output_size * sizeof(float));

    for (int i = 0 ; i < input_size[0] ; i++)
        input_buf[0][i] = 1.1 * i;

    for (int i = 0 ; i < input_size[1] ; i++)
        input_buf[1][i] = 2.2 * i;

    memset(output_buf, 0, output_size * sizeof(float));

    miopenCreate(&handle);

    miopenCreateTensorDescriptor(&outputDesc);
    miopenCreateTensorDescriptor(&inputDescs[0]);
    miopenCreateTensorDescriptor(&inputDescs[1]);

    miopenSetTensorDescriptor(inputDescs[0],
                              miopenFloat,
                              4,
                              input_shape[0],
                              nullptr);

    miopenSetTensorDescriptor(inputDescs[1],
                              miopenFloat,
                              4,
                              input_shape[1],
                              nullptr);

    miopenSetTensorDescriptor(outputDesc,
                              miopenFloat,
                              4,
                              output_shape,
                              nullptr);

    hipMalloc(&input_gpu_buf[0], input_size[0] * sizeof(float));
    hipMalloc(&input_gpu_buf[1], input_size[1] * sizeof(float));
    hipMalloc(&output_gpu_buf, output_size * sizeof(float));

    hipMemcpy(input_gpu_buf[0], input_buf[0], input_size[0] * sizeof(float), hipMemcpyHostToDevice);
    hipMemcpy(input_gpu_buf[1], input_buf[1], input_size[1] * sizeof(float), hipMemcpyHostToDevice);

    miopenCatForward(handle,
                     2,
                     inputDescs,
                     input_gpu_buf,
                     outputDesc,
                     output_gpu_buf,
                     dim);

    hipMemcpy(output_buf, output_gpu_buf, output_size * sizeof(float), hipMemcpyDeviceToHost);

    hipFree(input_gpu_buf[0]);
    hipFree(input_gpu_buf[1]);
    hipFree(output_gpu_buf);

    miopenDestroy(handle);

    input0 = input_buf[0];
    input1 = input_buf[1];
    output = output_buf;

    for (int i = 0 ; i < output_shape[0] ; i++)
    {
        for (int j = 0 ; j < input_size[0]/output_shape[0] ; j++)
        {
            if (*output != *input0)
                is_different = true;
            input0++;
            output++;
	}
        for (int j = 0 ; j < input_size[1]/output_shape[0] ; j++)
        {
            if (*output != *input1)
                is_different = true;
            input1++;
            output++;
        }
    }

    free(input_buf[0]);
    free(input_buf[1]);
    free(output_buf);

    if (is_different)
        printf("The input data and output data are distinct from each other.\n");
    else
        printf("The input data and output data are identical.\n");

    return 0
}
```

et16kr committed:
Adding content to the docs for cat

CAHEK7 commented:
A side note or further improvement step - concatenation is a type-agnostic operation - it doesn't matter do we want to concatenate f32 or f8 type, the only matters is a stride.

It means that it can be implemented like a `memcpy` primitive with the only single `void*` type, sizes and strides. In this case we need just one kernel for everything and under the hood it may use any sort of vectorization or coalescing techniques to achieve maximum memory throughput.

Another side note - it would be nice to calculate achieved memory throughput for this particular implementation.

JehandadKhan commented on include/miopen/miopen.h:
```diff
@@ -2557,6 +2558,35 @@ MIOPEN_EXPORT miopenStatus_t miopenLayerNormForward(miopenHandle_t handle,
+                                              const int32_t xCount,
+                                              const miopenTensorDescriptor_t* xDescs,
+                                              const void* const* xs,
+                                              const miopenTensorDescriptor_t& yDesc,
+                                              void* y,
```
Perhaps you meant:
```suggestion
                                              const miopenTensorDescriptor_t* yDesc,
                                              void* y,
```
unless I am missing something.

et16kr replied:
I referred to another API and removed the '&' symbols from the yDesc argument.

et16kr committed:
remove ref from y in miopenCatForward

et16kr committed:
remove dtype from cat forward kernel

et16kr committed:
cat forward vectorized prototype

et16kr committed:
update cat localsize for performance

et16kr committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
fix kernel cache key and error

et16kr commented:
1. I removed the dtype from kernel and changed it based on data size.
I didn't notice this. Thank you for the advice.
2. I modified it to copy 8 bytes at a time within a for loop.
3. In the sample, the new version is faster, so I did not maintain the non-vectorized approach.
4. I reduced the variety of keys for the kernel cache.
5. I changed the local size to 192 for performance.
6. While it becomes slower in small size tensors compared to the rocm pytorch, since small size tensors are already excluded by IsApplicable(), So I ignored it.


<details><summary>float16</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">




size | dim | model | direction | ROCm pytorch | prev MIOpen | Vectorized MIOpen | vs ROCm pytorch | vs non vectorized
-- | -- | -- | -- | -- | -- | -- | -- | --
[2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1560128 | 1553621 | 835279 | 1.87 | 1.86
[32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 17600 | 9938 | 8287 | 2.12 | 1.2
[32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 39200 | 25084 | 17048 | 2.3 | 1.47
[16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 128496 | 94151 | 57738 | 2.23 | 1.63
[64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 40608 | 31787 | 27597 | 1.47 | 1.15
[65536,16],[65536,351] | 1 | dlrm | fwd | 213136 | 240872 | 133964 | 1.59 | 1.8
[512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 341232 | 243147 | 159079 | 2.15 | 1.53
[256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 197744 | 166098 | 100177 | 1.97 | 1.66
[256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 82672 | 63840 | 40977 | 2.02 | 1.56
[2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 35696 | 24569 | 18759 | 1.9 | 1.31
[16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 34432 | 26880 | 20982 | 1.64 | 1.28
[3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 304816 | 116640 | 120072 | 2.54 | 0.97
[64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 56144 | 31147 | 21881 | 2.57 | 1.42
[12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1064288 | 346365 | 359140 | 2.96 | 0.96
[6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 44928 | 35147 | 25639 | 1.75 | 1.37
[255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 14608 | 13138 | 11757 | 1.24 | 1.12
[4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 529008 | 328747 | 246533 | 2.15 | 1.33
[256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 54944 | 32782 | 24149 | 2.28 | 1.36
[256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 96912 | 65244 | 46841 | 2.07 | 1.39
[2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 46528 | 24765 | 18146 | 2.56 | 1.36
[16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 40464 | 22756 | 17181 | 2.36 | 1.32
[4096,1312],[4096,1312] | 0 | rnnt | fwd | 58608 | 49742 | 33859 | 1.73 | 1.47
[9600,384],[128,384] | 0 | roberta | fwd | 27744 | 18578 | 14814 | 1.87 | 1.25
[256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 1672624 | 1017851 | 696801 | 2.4 | 1.46
[16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 68304 | 47698 | 31371 | 2.18 | 1.52
[1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 981168 | 840339 | 449353 | 2.18 | 1.87
[724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2123039 | 1905106 | 1111193 | 1.91 | 1.71



</body>

</html>

</details>

<details><summary>float32</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">


size | dim | model | direction | ROCm pytorch | prev MIOpen | vectorized MIOpen | vs ROCm pytorch | vs non vectorized
-- | -- | -- | -- | -- | -- | -- | -- | --
[2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1810432 | 1774453 | 1716809 | 1.05 | 1.03
[32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 19648 | 11716 | 11377 | 1.73 | 1.03
[32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 46544 | 29582 | 27228 | 1.71 | 1.09
[16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 148640 | 112427 | 113032 | 1.32 | 0.99
[64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 90480 | 43556 | 42869 | 2.11 | 1.02
[65536,16],[65536,351] | 1 | dlrm | fwd | 280464 | 273067 | 229805 | 1.22 | 1.19
[512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 430912 | 338454 | 303946 | 1.42 | 1.11
[256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 297472 | 178524 | 175487 | 1.7 | 1.02
[256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 99600 | 71502 | 70174 | 1.42 | 1.02
[2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 46944 | 32053 | 31465 | 1.49 | 1.02
[16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 65824 | 35307 | 34174 | 1.93 | 1.03
[3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 534320 | 215094 | 228138 | 2.34 | 0.94
[64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 57824 | 38756 | 38659 | 1.5 | 1
[12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1663424 | 648694 | 687715 | 2.42 | 0.94
[6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 90416 | 44942 | 43917 | 2.06 | 1.02
[255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 23600 | 15218 | 14400 | 1.64 | 1.06
[4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 974207 | 525370 | 502296 | 1.94 | 1.05
[256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 59920 | 49653 | 47200 | 1.27 | 1.05
[256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 107760 | 97387 | 92260 | 1.17 | 1.06
[2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 51552 | 32391 | 31570 | 1.63 | 1.03
[16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 44464 | 29013 | 28839 | 1.54 | 1.01
[4096,1312],[4096,1312] | 0 | rnnt | fwd | 96160 | 66205 | 64828 | 1.48 | 1.02
[9600,384],[128,384] | 0 | roberta | fwd | 43536 | 25724 | 25291 | 1.72 | 1.02
[256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 2807327 | 1580038 | 1475606 | 1.9 | 1.07
[16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 89424 | 58222 | 58240 | 1.54 | 1
[1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 1104256 | 900588 | 898705 | 1.23 | 1
[724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2332688 | 1915719 | 1848060 | 1.26 | 1.04




</body>

</html>


</details>

<details><summary>bfloat16</summary>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{white-space:normal;}
.xl66
	{color:#1F2328;
	font-family:Helvetica, sans-serif;
	mso-font-charset:0;
	white-space:normal;}
.xl67
	{color:#1F2328;
	white-space:normal;}
.xl68
	{mso-number-format:"0\.00_ ";
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->

</head>

<body link="#0563C1" vlink="#954F72">


size | dim | model | direction | ROCm pytorch | Prev MIOpen | vectorized MIOpen | vs ROCm pytorch | vs non vectorized
-- | -- | -- | -- | -- | -- | -- | -- | --
[2,32,128,128,128],[2,32,128,128,128] | 1 | 3dunet | fwd | 1561328 | 1550011 | 835141 | 1.87 | 1.86
[32,16,31,64],[32,16,1,64] | 2 | bart | fwd | 17504 | 9956 | 8292 | 2.11 | 1.2
[32,16,114,64],[32,16,1,64] | 2 | bart | fwd | 39360 | 25067 | 16993 | 2.32 | 1.48
[16,256,32,32],[16,256,32,32],[16,256,32,32],[16,256,32,32] | 1 | deeplabv3m | fwd | 148560 | 94951 | 57816 | 2.57 | 1.64
[64,1056,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7],[64,48,7,7] | 1 | densenet | fwd | 41760 | 31484 | 27779 | 1.5 | 1.13
[65536,16],[65536,351] | 1 | dlrm | fwd | 212672 | 240694 | 133899 | 1.59 | 1.8
[512,49,1024],[512,49,1024] | 2 | gnmt | fwd | 340272 | 242382 | 159007 | 2.14 | 1.52
[256,112,14,14],[256,288,14,14],[256,64,14,14],[256,64,14,14] | 1 | googlenet | fwd | 197952 | 166009 | 100188 | 1.98 | 1.66
[256,256,7,7],[256,320,7,7],[256,128,7,7],[256,128,7,7] | 1 | googlenet | fwd | 83264 | 64249 | 40960 | 2.03 | 1.57
[2,1024,768],[2,1024,768],[2,1024,768] | 0 | GPT | fwd | 35904 | 23733 | 18797 | 1.91 | 1.26
[16,192000,1],[16,48000,1],[16,12000,1],[16,3000,1],[16,780,1] | 1 | fasterrcnn | fwd | 34144 | 27076 | 20888 | 1.63 | 1.3
[3990480,1,1],[3990480,1,1],[3990480,1,1],[3990480,1,1] | 2 | fasterrcnn | fwd | 310192 | 116498 | 120232 | 2.58 | 0.97
[64,320,8,8],[64,192,8,8],[64,768,8,8] | 1 | inceptionv | fwd | 56192 | 31093 | 21953 | 2.56 | 1.42
[12277440,1],[12277440,1],[12277440,1],[12277440,1] | 1 | maskrcnn | fwd | 1060896 | 345974 | 358809 | 2.96 | 0.96
[6,182400,4],[6,45600,4],[6,11400,4],[6,2850,4],[6,741,4] | 1 | maskrcnn | fwd | 44672 | 35716 | 25611 | 1.74 | 1.39
[255780],[255780],[255780],[255780],[255780],[255780] | 0 | maskrcnn | fwd | 14528 | 13102 | 11763 | 1.24 | 1.11
[4,136800,91],[4,34200,91],[4,8550,91],[4,2223,91],[4,630,91] | 1 | retinanet | fwd | 526272 | 327218 | 246489 | 2.14 | 1.33
[256,232,7,7],[256,232,7,7] | 1 | shufflenet | fwd | 54240 | 32853 | 24143 | 2.25 | 1.36
[256,116,14,14],[256,116,14,14] | 1 | shufflenet | fwd | 96800 | 65582 | 47112 | 2.05 | 1.39
[2,192,80,80],[2,192,80,80] | 1 | yolor | fwd | 45392 | 24533 | 18124 | 2.5 | 1.35
[16,320,20,20],[16,320,20,20] | 1 | yolor | fwd | 40256 | 22382 | 16927 | 2.38 | 1.32
[4096,1312],[4096,1312] | 0 | rnnt | fwd | 57616 | 50027 | 34113 | 1.69 | 1.47
[9600,384],[128,384] | 0 | roberta | fwd | 27568 | 18293 | 14979 | 1.84 | 1.22
[256,81,5776],[256,81,2166],[256,81,600],[256,81,150],[256,81,36],[256,81,4] | 2 | ssd | fwd | 1672320 | 1002810 | 696432 | 2.4 | 1.44
[16,256,32,32],[16,128,32,32],[16,64,32,32],[16,64,32,32] | 1 | stdc | fwd | 66848 | 48284 | 31300 | 2.14 | 1.54
[1536,256,13,13],[1536,256,13,13] | 1 | squeezenet | fwd | 982031 | 841744 | 455769 | 2.15 | 1.85
[724,64,55,55],[724,64,55,55] | 1 | squeezenet | fwd | 2114175 | 1904376 | 1123219 | 1.88 | 1.7


</body>

</html>
</details>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{color:#1F2328;
	font-size:11.0pt;
	font-weight:700;
	font-family:"Segoe UI", sans-serif;
	mso-font-charset:0;
	text-align:center;
	white-space:normal;}
.xl66
	{color:#1F2328;
	font-size:11.0pt;
	font-family:"Segoe UI", sans-serif;
	mso-font-charset:0;
	white-space:normal;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:8.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"맑은 고딕", monospace;
	mso-font-charset:129;
	mso-char-type:none;
	display:none;}
-->
</head>

<body link="#0563C1" vlink="#954F72">


type | direction | avrage
-- | -- | --
float16 | fwd | 2.07
float | fwd | 1.62
bfloat16 | fwd | 2.08



</body>

</html>

CAHEK7 commented on src/include/miopen/cat/problem_description.hpp:
```diff
@@ -0,0 +1,135 @@
+
+private:
+    const TensorDescriptor* const* xDescs;
+    const TensorDescriptor& yDesc;
```
```suggestion
    const TensorDescriptor yDesc;
```
I overlooked it, but it is unsafe.
Check the following comments:
https://github.com/ROCm/MIOpen/pull/2671/files#r1466920140
https://github.com/ROCm/MIOpen/pull/2671/files#r1467003688

et16kr replied:
I understand the possibility of unintended mistakes. Thank you.

CAHEK7 reviewed (changes requested):
Unfortunately one more tine fix required.

Thanks @Vsevolod1983 to uncover that problem.

et16kr committed:
Update src/include/miopen/cat/problem_description.hpp

Change yDesc from reference to value in cat/problem_description.hpp

Co-authored-by: Alex Eremin <CAHEK7@yandex.ru>

JehandadKhan commented:
@et16kr  Can you please make this branch even with `develop`. @atamazov, @CAHEK7  and @Vsevolod1983 please finalize reviews.

et16kr committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
Merge branch 'develop' into impl_cat_forward

et16kr committed:
Fix code by clang-format in the merge develop into cat

et16kr committed:
Addressed review feedback: add [[maybe_unused]], init struct member variables, fix etc
